### PR TITLE
Add new contributing guidelines

### PR DESCRIPTION
**Pull request by qaisjp**
This pull request adds contributing guidelines that should make this project more suitable for greater numbers of collaborators.

I am looking for feedback on:
- [the "Contributing" guide](https://github.com/multitheftauto/mtasa-blue/blob/feature/new-contributing-guidelines/CONTRIBUTING.md) - these are rules
- [the "Code Review" guide](https://github.com/multitheftauto/mtasa-blue/wiki/Code-Review) - these are guidelines

This is opinionated, so if you disagree with anything, please suggest improvements.

----

~~**The below change has been abandoned.**~~ It _was_ abandoned in the hope to attract more reviews, but it didn't work. So I'm sticking to my guns. 

One major _point of contention_ is the final **Merging pull requests** section, specifically the part about enabling GitHub's _branch protection_.

This change means that, the following people that can currently commit directly to `master`, will now have to submit PRs: @Ligushka, @lopezloo, @patrikjuvonen and @Dutchman101.

It is not possible to disable branch protection for individual users without disabling branch protection for everyone. This is a GitHub limitation.

Please argue for or against -- and then we can bring on some more collaborators! [**Anyone can send a review by clicking "Review changes" here.**](https://github.com/multitheftauto/mtasa-blue/pull/1308/files)

### CODE REVIEW

**Review by qaisjp — COMMENTED** [review]
**For**

Generally, every change should be looked at by at least two pairs of eyes (including the author of the change), and I think enabling branch protection is a good step forward. Therefore I'm pledging to send pull requests ~~as much as possible~~ for all changes ([which I've already started doing](https://github.com/multitheftauto/mtasa-blue/pulls?q=is%3Apr+sort%3Aupdated-desc+author%3Aqaisjp+created%3A%3E%3D2019-04-01+)). It's also a good reason why repo admins should be using PRs - not everything there has been merged.

An argument against this is that we are slow with reviewing pull requests, and this will make it harder for collaborators to make changes. My take on this is that `more collaborators = more reviewers`, and that collaborators will feel like reviewing other PRs if their own PRs are waiting on being merged.

I also get the feeling that, without branch protection, we will be more hesitant to promote people to collaborator status.

Personally I'd be for enforcing it for admins too, but I haven't included this in the PR because this is less likely to be approved.

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md]:
```suggestion
**Branches are "topical" and should not be "personal" to each
```

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md:72]: One or two examples should be added here

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md]:
```suggestion
**Try to keep pull requests small — they should be about one thing.** you do multiple things
```

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md]:
```suggestion
The subject matter of your pull requests do not matter — we are more interested in,
```

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md]:
```suggestion
in one pull request, it's hard to review. If you're fixing stuff as you go, you might want
```

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md]:
```suggestion
leaving the pull request clean.
```

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md]:
```suggestion
including the previous commit-identifier SHA and a summarised commit message in
```

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md]:
```suggestion
or you fail to stick to the rules, your permissions will be revoked.
```

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md]:
```suggestion
and those people are therefore not required to send pull requests. Individual repository admins may,
```

File: CONTRIBUTING.md
Comment (StrixG) [inline CONTRIBUTING.md]:
```suggestion
**Try to keep pull requests small — they should be about one thing.** When you do multiple things
```

File: CONTRIBUTING.md
Comment (qaisjp) [reply CONTRIBUTING.md:72]: How is this now?

Comment (botder) [reply CONTRIBUTING.md:72]: Nice

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md]:
```suggestion
    telling us which pull request created this commit. There's no description.
```

File: CONTRIBUTING.md
Comment (botder) [inline CONTRIBUTING.md]:
```suggestion
make your pull request more likely to be merged in a timely fashion.
```

File: CONTRIBUTING.md
Comment (qaisjp) [inline CONTRIBUTING.md]: **Branch protection**: I am summarising comments here from a private channel until the relevant parties send their PR review:
- `-1` one collaborator is not in favour of losing their own direct commit access
- `+1` one collaborator does not mind losing their own direct commit access, and appears to be in support 
- `··` two inactive collaborators that would lose direct commit access, appear to abstain 
- `-1` one repo admin does not think existing permissions should change 
- `-1` one repo admin is not in favour of enforcing branch protection due to risk of "[not having] people reviewing pull requests on a timely basis" 
- `+2` two repo admins appear to be in support (myself included) 
- `··` the remaining six repo admins, three of which are inactive, appear to abstain 

So at this time of this comment, it sums up to `+3 -3`. To get things moving, I've **drafted** the following suggestion in case branch protection is unwanted.

Please send improvements to this suggestion, in the comments, in the form of a GitHub _suggested change_.

```suggestion
Before merging, pull requests _should_ have:
- Linux and Windows status checks passing, and
- 1 pull request review, or 2 reviews for large pull requests

These rules are not enforced via branch protection but, for the greater good,
please try and stick to this convention (... unless nobody is reviewing your PR).
```

**Comment by qaisjp** [comment]
I've abandoned the requirement for pull request reviews in the hope that it brings more review approvals.

Please send your reviews, @multitheftauto/blue-collaborators.

File: CONTRIBUTING.md
Comment (sbx320) [inline CONTRIBUTING.md]: Not sure if it's really neccessary for us to have a strict length limit. Those length rules are more commonly used for projects where people commonly use terminal based editors etc. We aren't following that rule at all right now either. I did some fancy bash scripting and out of the 100 most recent commits, 40 of them would violate that rule. Among the entire repository commit tree, 3687/7536 commits have a subject line above 50 characters.

And because I have that in a shell right now. Here's the number for more values:

For the past 2000 commits (to stay after our Git move, earlier SVN commits are much longer, 1196/7536 are above 90 chars total)
 
- 761/2000: Above 50 
- 410/2000: Above 60 
- 230/2000: Above 70 
- 130/2000: Above 80 
- 82/2000: Above 90

**Review by sbx320 — COMMENTED** [review]
The only thing partially missing from here is the decision between when to create a PR and when not to (for those with direct commit access). Likely not that relevant as it has been working well in the recent past.

File: CONTRIBUTING.md
Comment (qaisjp) [reply CONTRIBUTING.md]: I'm not rigid on this one. You're prolly right.

My only argument would be that GitHub collapses anything longer than 70. E.g. https://github.com/multitheftauto/mtasa-blue/commits/master?after=563f22bf40c6c613865e48fb4b6561e22d5b3874+0

So I'd suggest preferring 70 instead:

```suggestion
2. Limit the subject line to 70 characters (the [seven rules] say 50, but we think 70 is okay)
```

but if you prefer getting rid of it altogether, that's fine too

```suggestion
```

Comment (qaisjp) [reply CONTRIBUTING.md]: Or something midway that allows for more flexibility around that border

```suggestion
2. Limit the subject line to around 60-80 characters (the [seven rules] say 50, but we think ~70 is okay)
```

**Comment by qaisjp** [comment]
Also could add some stuff about how/when to use milestones

File: CONTRIBUTING.md
Comment (qaisjp) [reply CONTRIBUTING.md]: Any feedback on this @sbx320?

Comment (sbx320) [reply CONTRIBUTING.md]: I like the last one best, but either is fine

**Comment by qaisjp** [comment]
Milestones added here https://github.com/multitheftauto/mtasa-blue/wiki#milestones instead of in the contributing guide

**Comment by qaisjp** [comment]
Not sure if adding 933d24a was a good idea, it seemed to be ineffective (only seemed to attract a review from @sbx320).

image

**Review by Dutchman101 — APPROVED** [review]

**Comment by Dutchman101** [comment]
I intended to approve 933d24a only, not sure what went wrong

> Not sure if adding [933d24a](https://github.com/multitheftauto/mtasa-blue/commit/933d24abb9e3de8f50b29eaac74c1331aef37ad3) was a good idea, it seemed to be ineffective (only seemed to attract a review from @sbx320).

It's a good idea, in fact it was the outcome of an internal discussion, as you know.. (but others don't). Please don't try to force your original plans that you had to soften up in 933d24a - it will upset people, including myself, and you know exactly why.

Anyways, now you got 2 approval reviews on it.

image

The above should take the matter out of question and make a low amount of reviews sufficient

**Pull request merged**
